amethyst_renderer/src/lib.rs

```diff
@@ -0,0 +1,284 @@
+
+pub use cam::{Camera, Projection};
+pub use color::Rgba;
+pub use error::{Error, Result};
+pub use light::Light;
+pub use mesh::{Mesh, MeshBuilder};
+pub use mtl::{Material, MaterialBuilder};
+pub use pipe::{Pipeline, PipelineBuilder, Stage, Target};
+pub use scene::{LightIter, MeshIter, Scene};
+pub use tex::{Texture, TextureBuilder};
+pub use vertex::VertexFormat;
+
+use pipe::{ColorBuffer, DepthBuffer};
+use std::sync::Arc;
+use std::time::Duration;
+use types::{ColorFormat, DepthFormat, Encoder, Factory, Window};
+use winit::WindowBuilder;
+
+#[cfg(feature = "opengl")]
+use glutin::EventsLoop;
+#[cfg(not(feature = "opengl"))]
+use winit::EventsLoop;
+
+pub mod light;
+pub mod pass;
+pub mod prelude;
+pub mod pipe;
+pub mod vertex;
+
+mod cam;
+mod color;
+mod error;
+mod mesh;
+mod mtl;
+mod scene;
+mod tex;
+mod types;
+
+/// Generic renderer.
+pub struct Renderer {
+    device: types::Device,
+    encoders: Vec<Encoder>,
+    factory: Factory,
+    main_target: Arc<Target>,
+    pool: rayon::ThreadPool,
+    window: Window,
+}
+
+impl Renderer {
+    /// Creates a new renderer with default window settings.
+    pub fn new(el: &EventsLoop) -> Result<Renderer> {
+        let wb = WindowBuilder::new().with_title("Amethyst");
+        Renderer::from_winit_builder(wb, el)
+    }
+
+    /// Creates a new renderer with the given `winit::WindowBuilder`.
+    pub fn from_winit_builder(wb: WindowBuilder, el: &EventsLoop) -> Result<Renderer> {
+        let Backend(dev, mut fac, main, win) = init_backend(wb, el)?;
+
+        let num_cores = num_cpus::get();
+        let encoders = (0..num_cores)
+            .map(|_| fac.create_command_buffer().into())
+            .collect();
+
+        let cfg = rayon::Configuration::new().num_threads(num_cores);
+        let pool = rayon::ThreadPool::new(cfg)
+            .map_err(|e| Error::PoolCreation(e))?;
+
+        Ok(Renderer {
+            device: dev,
+            encoders: encoders,
+            factory: fac,
+            main_target: Arc::new(main),
+            pool: pool,
+            window: win,
+        })
+    }
+
+    /// Builds a new mesh from the given vertices.
+    pub fn create_mesh(&mut self, mb: MeshBuilder) -> Result<Mesh> {
+        mb.build(&mut self.factory)
+    }
+
+    /// Builds a new renderer pipeline.
+    pub fn create_pipe(&mut self, pb: PipelineBuilder) -> Result<Pipeline> {
+        pb.build(&mut self.factory, &self.main_target)
+    }
+
+    /// Builds a new texture resource.
+    pub fn create_texture(&mut self, tb: TextureBuilder) -> Result<Texture> {
+        tb.build(&mut self.factory)
+    }
+
+    /// Draws a scene with the given pipeline.
+    pub fn draw(&mut self, scene: &Scene, pipe: &Pipeline, _delta: Duration) {
+        use gfx::Device;
+        use rayon::prelude::*;
+
+        {
+            let encoders = self.encoders.as_mut_slice();
+            self.pool.install(|| {
+                pipe.stages()
+                    .par_iter()
```

Comment:
I think if we want to use rayon then `.par_iter` should be called on a `Vec` of `Fragment`s. Here is a code example illustrating how this might work:
```rust
// Warning: Pseudo code - probably won't compile.
// encoders: Vec<Encoder>;
assert!(encoders.length() == num_cpus);
let fragments_len = fragments.length();
let repeating_encoders = encoders.iter().cycle().take(fragments_len);
for stage in pipe.stages() {
    if stage.is_enabled() {
        self.pool.install(|| {
            fragments
                .par_iter()
                .zip(repeating_encoders)
                .for_each(|fragment, enc| stage.apply(fragment, enc, scene));
            // scene contains lights and other things that are the same for all fragments
        }
    }
}
```